### PyTorch

<img src=https://upload.wikimedia.org/wikipedia/commons/thumb/c/c6/PyTorch_logo_black.svg/1920px-PyTorch_logo_black.png width="700" style="background:white">

- https://pytorch.org/ 공식사이트
- https://pytorch.kr/ 한글 공식사이트
- https://tutorials.pytorch.kr/ 튜토리얼만 한글화 사이트

#### PyTorch 설치
- CPU 설치
    - 간단하지만, 그래픽카드 가속 사용할 수 없음
    - 기본 파이썬 환경 설치

- GPU 설치
    - 가상환경 생성, 설치

#### CPU 설치
- https://pytorch.org/get-started/locally/ 나와있는 Note에서 해당 플랫폼 종류 선택, 명령어 복사 실행

    ```python
    !pip install torch torchvision torchaudio
    ```

In [2]:
!pip install torch torchvision torchaudio

  Using cached torch-2.4.0-cp311-cp311-win_amd64.whl.metadata (27 kB)
  Using cached torchvision-0.19.0-1-cp311-cp311-win_amd64.whl.metadata (6.1 kB)
  Using cached torchaudio-2.4.0-cp311-cp311-win_amd64.whl.metadata (6.4 kB)
  Using cached sympy-1.13.1-py3-none-any.whl.metadata (12 kB)
  Using cached networkx-3.3-py3-none-any.whl.metadata (5.1 kB)
  Using cached fsspec-2024.6.1-py3-none-any.whl.metadata (11 kB)
  Using cached mpmath-1.3.0-py3-none-any.whl.metadata (8.6 kB)
   ---------------------------------------- 0.0/197.9 MB ? eta -:--:--
   ---------------------------------------- 0.8/197.9 MB 3.7 MB/s eta 0:00:53
   ---------------------------------------- 1.3/197.9 MB 3.7 MB/s eta 0:00:53
   ---------------------------------------- 2.1/197.9 MB 3.5 MB/s eta 0:00:57
    --------------------------------------- 2.9/197.9 MB 3.5 MB/s eta 0:00:56
    --------------------------------------- 3.7/197.9 MB 3.6 MB/s eta 0:00:55
    --------------------------------------- 4.5/197.9 MB 3.6

In [3]:
import torch

In [4]:
## 2.4.0+cpu = CPU 용 설치
torch.__version__

'2.4.0+cpu'

#### PyTorch GPU 설치
- NVIDIA GPU에서만 가능, CUDA로 되어있기 때문
- GPU설치가 가능한지 먼저 확인

```python
!nvidia-smi
```

In [5]:
!nvidia-smi

Tue Aug  6 12:31:41 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 536.23                 Driver Version: 536.23       CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce GTX 1650      WDDM  | 00000000:01:00.0  On |                  N/A |
| 50%   32C    P8               9W /  75W |    766MiB /  4096MiB |      6%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

##### CUDA Version 12.2
- CUDA(Computer Unified Device Architecture) : NVIDIA에서 그래픽카드 가속을 위해 만든 기술
- https://ko.wikipedia.org/wiki/CUDA

##### CUDA 설치
- CUDA Toolkit 12.2
    - https://developer.nvidia.com/cuda-12-2-2-download-archive?target_os=Windows&target_arch=x86_64&target_version=11
- cuDNN(Deep Neural Network)
    - CUDA Toolkit에 맞는 버전확인
    - https://developer.nvidia.com/rdp/cudnn-archive 확인 후 다운로드(단, 로그인 필수)
    - 압축을 풀고 (CUDA 경로에 복사/붙여넣기)

#### GPU용 파이썬 가상환경 만들기
- GPU PyTorch는 설치로그를 봐야하기 때문에 명령 프롬프트에서 진행

- 터미널 오픈
    - pip list로 virtualenv 확인
    - 없으면 pip install virtualenv 설치
    
- 가상환경 활성화
    - cd gpu_env\Scripts\activate
    - pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121

#### 파이토치 GPU 확인

In [1]:
import torch

In [2]:
## 2.4.0+cu121 --> CUDA 12.1 버전의 PyTorch
torch.__version__

'2.4.0+cu121'

In [3]:
## CUDA 쓸 수 있는지 확인
torch.cuda.is_available()

True

In [4]:
## 설치된 CUDA가 동작하는 그래픽카드 정보확인
torch.cuda.get_device_name(0)

'NVIDIA GeForce GTX 1650'

#### 이제부터 GPU 가속으로 파이토치(텐서플로 가능) 딥러닝을 할 수 있음